be/src/storage/persistent_index.cpp

```diff
@@ -922,11 +1354,561 @@ StatusOr<std::unique_ptr<MutableIndex>> MutableIndex::create(size_t key_size, co
+
+bool ShardByLengthMutableIndex::dump(phmap::BinaryOutputArchive& ar_out) {
+    for (const auto& shard : _shards) {
+        if (!shard->dump(ar_out)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+Status ShardByLengthMutableIndex::commit(MutableIndexMetaPB* meta, const EditVersion& version, const CommitType& type) {
+    std::shared_ptr<FileSystem> fs;
+    ASSIGN_OR_RETURN(fs, FileSystem::CreateSharedFromString(_path));
+    switch (type) {
+    case kFlush: {
+        // create a new empty _l0 file because all data in _l0 has write into _l1 files
+        std::string file_name = get_l0_index_file_name(_path, version);
+        WritableFileOptions wblock_opts;
+        wblock_opts.mode = FileSystem::CREATE_OR_OPEN_WITH_TRUNCATE;
+        ASSIGN_OR_RETURN(auto wfile, fs->new_writable_file(wblock_opts, file_name));
+        DeferOp close_block([&wfile] {
+            if (wfile) {
+                wfile->close();
+            }
+        });
+        meta->clear_wals();
+        IndexSnapshotMetaPB* snapshot = meta->mutable_snapshot();
+        version.to_pb(snapshot->mutable_version());
+        PagePointerPB* data = snapshot->mutable_data();
+        // create a new empty _l0 file, set _offset to 0
+        data->set_offset(0);
+        data->set_size(0);
+        meta->set_format_version(PERSISTENT_INDEX_VERSION_2);
+        _offset = 0;
+        std::fill(_page_sizes.begin(), _page_sizes.end(), 0);
+        break;
+    }
+    case kSnapshot: {
+        std::string file_name = get_l0_index_file_name(_path, version);
+        // be maybe crash after create index file during last commit
+        // so we delete expired index file first to make sure no garbage left
+        FileSystem::Default()->delete_file(file_name);
+        size_t snapshot_size = dump_bound();
+        phmap::BinaryOutputArchive ar_out(file_name.data());
+        if (!dump(ar_out)) {
+            std::string err_msg = strings::Substitute("faile to dump snapshot to file $0", file_name);
+            LOG(WARNING) << err_msg;
+            return Status::InternalError(err_msg);
+        }
+        // dump snapshot success, set _index_file to new snapshot file
+        WritableFileOptions wblock_opts;
+        wblock_opts.mode = FileSystem::MUST_EXIST;
+        ASSIGN_OR_RETURN(_index_file, fs->new_writable_file(wblock_opts, file_name));
+        meta->clear_wals();
+        IndexSnapshotMetaPB* snapshot = meta->mutable_snapshot();
+        version.to_pb(snapshot->mutable_version());
+        PagePointerPB* data = snapshot->mutable_data();
+        data->set_offset(0);
+        data->set_size(snapshot_size);
+        meta->set_format_version(PERSISTENT_INDEX_VERSION_2);
+        _offset = snapshot_size;
+        std::fill(_page_sizes.begin(), _page_sizes.end(), 0);
+        break;
+    }
+    case kAppendWAL: {
+        for (const auto page_size : _page_sizes) {
+            if (page_size != 0) {
+                IndexWalMetaPB* wal_pb = meta->add_wals();
```

Comment:
I think we should only generate one WALPB of one version. And this wal page contains the contents of all shard in L0